### ISSUE TITLE
Android crash  'Unable to post metrics; will retry'

### ISSUE THREAD

**Issue by PiotrD-MR** [boxcast/boxcast-sdk-js#8]
I have a problem with crash on Android release caused by boxcast-sdk. This happens right after calling `this.analytics.generateVideoEventProps().onBuffer(data)`.
On debug mode there is only console.error - no crash. 
I have been using this library for 2 months and never before have I problems like that. 

```
2019-09-26 15:19:53.982 8788-8896/? E/ReactNativeJS: 'Unable to post metrics; will retry', { [Error: Request failed with status code 400]
      config: 
       { adapter: [Function],
         transformRequest: { '0': [Function] },
         transformResponse: { '0': [Function] },
         timeout: 0,
         xsrfCookieName: 'XSRF-TOKEN',
         xsrfHeaderName: 'X-XSRF-TOKEN',
         maxContentLength: -1,
         validateStatus: [Function: validateStatus],
         headers: 
          { Accept: 'application/json, text/plain, */*',
            'Content-Type': 'application/json;charset=utf-8' },
         method: 'post',
         url: 'https://metrics.boxcast.com/player/interaction',
         data: '{"view_id":"F2C61E5636D3A8F5BFA210F17","viewer_id":"6BD3ABA8DB4EE9C94F4F9D22E","channel_id":"xnzywc34gmxf4f1cyoru","is_live":false,"broadcast_id":"u7eddjgzhaj11fjnyvx0","timestamp":"2019-09-26T13:19:53.273Z","hour_of_day":15,"day_of_week":4,"action":"buffer","duration":0,"duration_buffering":2}' },
      request: 
       { UNSENT: 0,
         OPENED: 1,
         HEADERS_RECEIVED: 2,
         LOADING: 3,
         DONE: 4,
         readyState: 4,
         status: 400,
         timeout: 0,
         withCredentials: true,
         upload: {},
         _aborted: false,
         _hasError: false,
         _method: 'POST',
         _response: '{"error":"bad_request","error_description":"The request was invalid."}',
         _url: 'https://metrics.boxcast.com/player/interaction',
         _timedOut: false,
         _trackingName: 'unknown',
         _incrementalEvents: false,
         responseHeaders: 
          { 'x-content-type-options': 'nosniff',
            vary: 'Origin',
            'userve-route': '/player/interaction',
            'content-length': '70',
            'content-type': 'application/json; charset=utf-8',
            date: 'Thu, 26 Sep 2019 13:19:54 GMT' },
         _requestId: null,
         _cachedResponse: undefined,
         _headers: 
          { accept: 'application/json, text/plain, */*',
            'content-type': 'application/json;charset=utf-8' },
         _responseType: '',
         _sent: true,
         _lowerCaseResponseHeaders: 
          { 'x-content-type-options': 'nosniff',
            vary: 'Origin',
            'userve-route': '/player/interaction',
            'content-length': '70',
            'content-type': 'application/json; charset=utf-8',
            date: 'Thu, 26 Sep 2019 13:19:54 GMT' },
         _subscriptions: [],
         responseURL: 'https://metrics.boxcast.com/player/interaction' },
      response: 
       { data: 
          { error: 'bad_request',
            error_description: 'The request was invalid.' },
         status: 400,
         statusText: undefined,
         headers: 
          { 'x-content-type-options': 'nosniff',
            vary: 'Origin',
            'userve-route': '/player/interaction',
            'content-length': '70',
            'content-type': 'application/json; charset=utf-8',
            date: 'Thu, 26 Sep 2019 13:19:54 GMT' },
         config: 
          { adapter: [Function],
            transformRequest: { '0': [Function] },
            transformResponse: { '0': [Function] },
            timeout: 0,
            xsrfCookieName: 'XSRF-TOKEN',
            xsrfHeaderName: 'X-XSRF-TOKEN',
            maxContentLength: -1,
            validateStatus: [Function: validateStatus],
            headers: 
             { Accept: 'application/json, text/plain, */*',
               'Content-Type': 'application/json;charset=utf-8' },
            method: 'post',
            url: 'https://metrics.boxcast.com/player/interaction',
            data: '{"view_id":"F2C61E5636D3A8F5BFA210F17","viewer_id":"6BD3ABA8DB4EE9C94F4F9D22E","channel_id":"xnzywc34gmxf4f1cyoru","is_live":false,"broadcast_id":"u7eddjgzhaj11fjnyvx0","timestamp":"2019-09-26T13:19:53.273Z","hour_of_day":15,"day_of_week":4,"action":"buffer","duration":0,"duration_buffering":2}' },
         request: 
          { UNSENT: 0,
            OPENED: 1,
            HEADERS_RECEIVED: 2,
            LOADING: 3,
            DONE: 4,
            readyState: 4,
            status: 400,
            timeout: 0,
            withCredentials: true,
            upload: {},
            _aborted: false,
            _hasError: false,
            _method: 'POST',
            _response: '{"error":"bad_request","error_description":"The request was invalid."}',
            _url: 'https://metrics.boxcast.com/player/interaction',
           
2019-09-26 15:19:54.009 8788-8896/? E/ReactNativeJS: 'Unable to post metrics; will retry', { [Error: Request failed with status code 400]
      config: 
       { adapter: [Function],
         transformRequest: { '0': [Function] },
         transformResponse: { '0': [Function] },
         timeout: 0,
         xsrfCookieName: 'XSRF-TOKEN',
         xsrfHeaderName: 'X-XSRF-TOKEN',
         maxContentLength: -1,
         validateStatus: [Function: validateStatus],
         headers: 
          { Accept: 'application/json, text/plain, */*',
            'Content-Type': 'application/json;charset=utf-8' },
         method: 'post',
         url: 'https://metrics.boxcast.com/player/interaction',
         data: '{"view_id":"F2C61E5636D3A8F5BFA210F17","viewer_id":"6BD3ABA8DB4EE9C94F4F9D22E","channel_id":"xnzywc34gmxf4f1cyoru","is_live":false,"broadcast_id":"u7eddjgzhaj11fjnyvx0","host":"","os":"null","browser_name":"React Native","browser_version":"1.0","player_version":"AirDance v1.0","timestamp":"2019-09-26T13:19:53.270Z","hour_of_day":15,"day_of_week":4,"action":"setup","duration":0,"duration_buffering":1}' },
      request: 
       { UNSENT: 0,
         OPENED: 1,
         HEADERS_RECEIVED: 2,
         LOADING: 3,
         DONE: 4,
         readyState: 4,
         status: 400,
         timeout: 0,
         withCredentials: true,
         upload: {},
         _aborted: false,
         _hasError: false,
         _method: 'POST',
         _response: '{"error":"bad_request","error_description":"The request was invalid."}',
         _url: 'https://metrics.boxcast.com/player/interaction',
         _timedOut: false,
         _trackingName: 'unknown',
         _incrementalEvents: false,
         responseHeaders: 
          { 'x-content-type-options': 'nosniff',
            vary: 'Origin',
            'userve-route': '/player/interaction',
            'content-length': '70',
            'content-type': 'application/json; charset=utf-8',
            date: 'Thu, 26 Sep 2019 13:19:54 GMT' },
         _requestId: null,
         _cachedResponse: undefined,
         _headers: 
          { accept: 'application/json, text/plain, */*',
            'content-type': 'application/json;charset=utf-8' },
         _responseType: '',
         _sent: true,
         _lowerCaseResponseHeaders: 
          { 'x-content-type-options': 'nosniff',
            vary: 'Origin',
            'userve-route': '/player/interaction',
            'content-length': '70',
            'content-type': 'application/json; charset=utf-8',
            date: 'Thu, 26 Sep 2019 13:19:54 GMT' },
         _subscriptions: [],
         responseURL: 'https://metrics.boxcast.com/player/interaction' },
      response: 
       { data: 
          { error: 'bad_request',
            error_description: 'The request was invalid.' },
         status: 400,
         statusText: undefined,
         headers: 
          { 'x-content-type-options': 'nosniff',
            vary: 'Origin',
            'userve-route': '/player/interaction',
            'content-length': '70',
            'content-type': 'application/json; charset=utf-8',
            date: 'Thu, 26 Sep 2019 13:19:54 GMT' },
         config: 
          { adapter: [Function],
            transformRequest: { '0': [Function] },
            transformResponse: { '0': [Function] },
            timeout: 0,
            xsrfCookieName: 'XSRF-TOKEN',
            xsrfHeaderName: 'X-XSRF-TOKEN',
            maxContentLength: -1,
            validateStatus: [Function: validateStatus],
            headers: 
             { Accept: 'application/json, text/plain, */*',
               'Content-Type': 'application/json;charset=utf-8' },
            method: 'post',
            url: 'https://metrics.boxcast.com/player/interaction',
            data: '{"view_id":"F2C61E5636D3A8F5BFA210F17","viewer_id":"6BD3ABA8DB4EE9C94F4F9D22E","channel_id":"xnzywc34gmxf4f1cyoru","is_live":false,"broadcast_id":"u7eddjgzhaj11fjnyvx0","host":"","os":"null","browser_name":"React Native","browser_version":"1.0","player_version":"AirDance v1.0","timestamp":"2019-09-26T13:19:53.270Z","hour_of_day":15,"day_of_week":4,"action":"setup","duration":0,"duration_buffering":1}' },
         request: 
          { UNSENT: 0,
            OPENED: 1,
            HEADERS_RECEIVED: 2,
            LOADING: 3,
            DONE: 4,
            readyState: 4,
            status: 400,
            timeout: 0,
            withCredentials: true,
            upload: {},
            _aborted: false,
    
    --------- beginning of crash
2019-09-26 15:19:54.106 8788-8897/? E/AndroidRuntime: FATAL EXCEPTION: mqt_native_modules
    Process: com.airdance, PID: 8788
    com.facebook.react.bridge.NoSuchKeyException: lineNumber
        at com.facebook.react.bridge.ReadableNativeMap.getValue(ReadableNativeMap.java:124)
        at com.facebook.react.bridge.ReadableNativeMap.getValue(ReadableNativeMap.java:128)
        at com.facebook.react.bridge.ReadableNativeMap.getInt(ReadableNativeMap.java:182)
        at com.facebook.react.util.JSStackTrace.format(JSStackTrace.java:28)
        at com.facebook.react.modules.core.ExceptionsManagerModule.reportSoftException(ExceptionsManagerModule.java:46)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:372)
        at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:158)
        at com.facebook.react.bridge.queue.NativeRunnable.run(Native Method)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.java:29)
        at android.os.Looper.loop(Looper.java:201)
        at com.facebook.react.bridge.queue.MessageQueueThreadImpl$4.run(MessageQueueThreadImpl.java:232)
        at java.lang.Thread.run(Thread.java:764)
```

**Comment by bikegriffith** [comment]
Thank you for the report.  Did you recently update your version of React Native?  Can you share which exact versions you are using of libraries and if this was in the simulator or on a real device?

The actual fatal exception points to something happening in the react bridge, but the line of code in question in the boxcast-sdk-js is doing nothing out of the ordinary (simply logging an HTTP 400 in pure-js code, nothing native or exotic).

```
FATAL EXCEPTION: mqt_native_modules
    Process: com.airdance, PID: 8788
    com.facebook.react.bridge.NoSuchKeyException: lineNumber
```

**Comment by PiotrD-MR** [comment]
Thank you for quick answer. I'm using:
      react: 16.8.3 
      react-native: 0.59.2
      boxcast-sdk-js: 1.3.1

No, I haven't changed any version recently. I have checked that on two real devices and on emulator. I suppose this crash is caused by boxcast-sdk because after commenting any analytics requests everything works fine.

**Comment by bikegriffith** [comment]
It appears as though the HTTP post is missing an `account_id`, and our server is raising the HTTP 400 when trying to look it up. We will look into the server side, but I'm curious why the account_id is missing.

Can you confirm, what arguments are you passing into the `analytics.attach(...)` method?  There should be a `broadcast` object, and that broadcast object should contain an `account_id`.

**Comment by bikegriffith** [comment]
@PiotrD-MR I published a boxcast-sdk-js v1.3.3, which simplifies the log handling and may resolve the crash.  We will also be releasing an update very soon to the metrics.boxcast.com cluster that will avoid the HTTP 400 you were seeing (finalizing testing on the fix right now).

Could you try to upgrade to boxcast-sdk-js v1.3.3. to see if that fixes your issue in advance of the server release?

**Comment by PiotrD-MR** [comment]
Yeah, everything works great now, thanks! :)
